src/cashlink.ts

```diff
@@ -1,76 +1,154 @@
-const { NumberUtils, SerialBuffer, BufferUtils, KeyPair, PrivateKey } = require('@nimiq/core');
-const { Utf8Tools } = require('@nimiq/utils');
-const { CashlinkTheme } = require('@nimiq/hub-api');
-
-const CashlinkExtraData = {
-    FUNDING:  new Uint8Array([0, 130, 128, 146, 135]), // 'CASH'.split('').map(c => c.charCodeAt(0) + 63)
-    CLAIMING: new Uint8Array([0, 139, 136, 141, 138]), // 'LINK'.split('').map(c => c.charCodeAt(0) + 63)
-};
-
-class Cashlink {
-    static parse(str) {
-        let [ baseUrl, hash ] = str.split('#');
-        hash = hash.replace(/~/g, '')
+/**
+ * Nimiq Cashlink Core Implementation
+ * Handles the creation, serialization, and management of Nimiq cashlinks.
+ * 
+ * Features:
+ * - Create and parse cashlinks
+ * - Handle message encoding/decoding
+ * - Support multiple themes
+ * - Manage serialization format
+ * - Handle key pair management
+ * 
+ * The cashlink class is the core component for cashlink generation and handling.
+ */
+
+import { NumberUtils, SerialBuffer, BufferUtils, KeyPair, PrivateKey, Address } from '@nimiq/core';
+
+/**
+ * Available themes for cashlink customization
+ */
+export enum CashlinkTheme {
+    UNSPECIFIED = 0,
+    STANDARD = 1,
+    CHRISTMAS = 2,
+    LUNAR_NEW_YEAR = 3,
+    EASTER = 4,
+    GENERIC = 5,
+    BIRTHDAY = 6
+}
+
+/**
+ * Extra data fields for cashlink transactions
+ */
+export const CashlinkExtraData = {
+    FUNDING: new Uint8Array([0, 130, 128, 146, 135]),
+    CLAIMING: new Uint8Array([0, 139, 136, 141, 138]),
+} as const;
+
+/**
+ * Core cashlink class for creating and managing Nimiq cashlinks
+ * Handles the creation, modification, and serialization of cashlinks
+ */
+export class Cashlink {
+    private _baseUrl: string;
+    private _keyPair: KeyPair;
+    private _value: number;
+    private _messageBytes: Uint8Array;
+    private _theme: number;
+    private static _textEncoder = new TextEncoder();
+    private static _textDecoder = new TextDecoder();
+
+    /**
+     * Parse a cashlink from its string representation
+     * @param str - Cashlink URL string to parse
+     * @returns New Cashlink instance
+     */
+    static parse(str: string): Cashlink {
+        const [baseUrl, rawHash] = str.split('#');
+        const hash = rawHash.replace(/~/g, '')
             .replace(/=*$/, (match) => new Array(match.length).fill('.').join(''));
         const buf = BufferUtils.fromBase64Url(hash);
-        const keyPair = KeyPair.derive(PrivateKey.unserialize(buf));
+        
+        // Ensure we preserve the /cashlink/ part in the URL if it exists
+        const normalizedBaseUrl = baseUrl.endsWith('/') ? baseUrl.slice(0, -1) : baseUrl;
+        const finalBaseUrl = normalizedBaseUrl.includes('/cashlink') ? 
+            normalizedBaseUrl : 
+            `${normalizedBaseUrl}/cashlink`;
```

Comment:
- Why is this necessary?
- The comment is wrong. This snippet does not only ensure the /cashlink/ is preserved, but also adds it if it's not present. Additionally, the trailing / is in fact not preserved or added, but removed.
- A better place for such a check would be the constructor, such that this invariant is also enforced for Cashlinks that are not created via the `parse` method.
- If the snippet is in fact needed (which I don't think), it can be simplified, under the assumption that the `/cashlink` part is supposed to be at the end of the URL.
```suggestion
        // Ensure the URL ends with /cashlink
        const normalizedBaseUrl = baseUrl.replace(/(?:\/cashlink)?\/?$/, '/cashlink');
```